---
## Hausaufgabe 2
---

### Achtung: Überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   25
---
---
Aufgabe 1: 10 <br>
<br>
Aufgabe 2: 4 <br>
<br>
Aufgabe 3: 11 <br>

#### Aufgabe 1 - First Things First (10 Punkte)

##### a.) (3 Punkte)

Importieren Sie zunächst alle Klassen und Objekte aus den mit gelieferten .py-Dateien. Diese kennen Sie bereits aus dem Seminar. Erzeugen Sie anschließend das Objekt **data** als Instanz der Klasse *InputData*. Als Inputdaten dienen Ihnen die Daten aus der mitgelieferten .json-Datei (VFR10_10_10_SIST.json). 

Für diese Beispieldaten wurde Ihnen aus der Fertigung die Reihenfolge 7-6-8-10-2-9-1-4-3-5 empfohlen. Erzeugen Sie hierfür ein Objekt der Klasse *Solution* und benennen Sie es **recommendedSolution**. Wie hoch ist der Makespan dieser Lösung, wenn Rüstzeiten vernachlässigt werden? Geben Sie *Solution* aus.


In [6]:
list1 = [0]*3
print(list1)

[0, 0, 0]


In [1]:
# Schreiben Sie hier Ihren Programmcode
## 1 Punkt für Import der Module 
from InputData import *
from OutputData import *
from EvaluationLogic import *

## 2 Punkte für Einladen und Testen
data = InputData('VFR10_10_10_SIST.json')
perm_list = [7, 6, 8, 10, 2, 9, 1, 4, 3, 5]

perm = [x - 1 for x in perm_list]

recommendedSolution = Solution(data.InputJobs, perm)
EvaluationLogic().DefineStartEnd(recommendedSolution)
print(recommendedSolution)

The permutation [6, 5, 7, 9, 1, 8, 0, 3, 2, 4] results in a Makespan of 1261


In [7]:
from InputData import *
from OutputData import *
from EvaluationLogic import *

data = InputData('VFR10_10_10_SIST.json')

permList = [7,6,8,10,2,9,1,4,3,5]

perm = [x-1 for x in permList]

recommendedSolution = Solution(data.InputJobs, perm)
print(recommendedSolution)

EvaluationLogic().DefineStartEnd(recommendedSolution)

print(recommendedSolution)


The permutation [6, 5, 7, 9, 1, 8, 0, 3, 2, 4] results in a Makespan of -1
The permutation [6, 5, 7, 9, 1, 8, 0, 3, 2, 4] results in a Makespan of 1261


Erwarteter Output:

    The permutation [6, 5, 7, 9, 1, 8, 0, 3, 2, 4] results in a Makespan of 1261

##### b.) (2 Punkte)

Erstellen Sie ein weiteres Objekt **simpleSolution** der Klasse *Solution*, bei dem die Aufträge nach der FCFS-Regel eingeplant werden. Berechnen Sie das Delta zwischen den beiden Lösungen **recommendedSolution** and **simpleSolution** und geben Sie dieses aus.


In [2]:
# Schreiben Sie hier Ihren Programmcode
# 1 Punkt: Erstellung der Solution & Aufruf Evaluation
simpleSolution = Solution(data.InputJobs, range(len(data.InputJobs)))
EvaluationLogic().DefineStartEnd(simpleSolution)
# 1 Punkt: Berechnung Delta und Ausgabe
delta = recommendedSolution.Makespan - simpleSolution.Makespan
print(f'The objective delta is {delta}')

The objective delta is -11


In [8]:
simpleSolution = Solution(data.InputJobs, [x for x in range(len(data.InputJobs))])
EvaluationLogic().DefineStartEnd(simpleSolution)

print(f'The objective delta is {simpleSolution.Makespan - recommendedSolution.Makespan}')

The objective delta is 11


Erwarteter Output:

    The objective delta is -11.

##### c.) (5 Punkte)

Schreiben Sie eine Methode *CalcuateTardyJobs*, welche die **Anzahl** der verspäteten Aufträge für eine Lösung errechnet. Fügen Sie diese Methode anschließend der Klasse *EvaluationLogic* hinzu und bestimmen Sie die Anzahl der verspäteten Aufträge für die beiden Lösungen **recommendedSolution** und **simpleSolution**. 

In [3]:
# Schreiben Sie hier Ihren Programmcode
# Schreiben der Funktion 2,5 Punkte
def CalcuateTardyJobs(solution):
    late_jobs = []
    for job in solution.OutputJobs.values():
        finish = job.EndTimes[-1]

        if finish > job.DueDate:
            late_jobs.append(job)
    return len(late_jobs)

# Hinzufügen zur Klasse 0,5 Punkte
setattr(EvaluationLogic, 'CalculateTardyJobs', CalcuateTardyJobs)

# Aufruf der neuen Funktion 1 Punkt
# Erstellung der Ausgabe 1 Punkt
print(f'Tardy jobs of simpleSolution: {CalcuateTardyJobs(simpleSolution)}')
print(f'Tardy jobs of recommendedSolution: {CalcuateTardyJobs(recommendedSolution)}')


Tardy jobs of simpleSolution: 7
Tardy jobs of recommendedSolution: 8


In [16]:
def CalculateTardyJobs(solution):
    delayed = 0
    for id in solution.OutputJobs:  ### 对字典的key进行遍历，比较麻烦
        if solution.OutputJobs[id].EndTimes[-1] > solution.OutputJobs[id].DueDate:
            delayed += 1

    return delayed
setattr(EvaluationLogic, 'CalculateTardyJobs', CalcuateTardyJobs)

print(f'Tardy jobs of simpleSolution: {CalcuateTardyJobs(simpleSolution)}')
print(f'Tardy jobs of recommendedSolution: {CalcuateTardyJobs(recommendedSolution)}')


Tardy jobs of simpleSolution: 7
Tardy jobs of recommendedSolution: 8


In [13]:
def CalculateTardyJobs(solution):
    delayed = 0
    for job in solution.OutputJobs.values():
        finish = job.EndTImes[-1]

        if finish > job.DueDate:
            delayed += 1
    return delayed

print(f'Tardy jobs of simpleSolution: {CalcuateTardyJobs(simpleSolution)}')
print(f'Tardy jobs of recommendedSolution: {CalcuateTardyJobs(recommendedSolution)}')

Tardy jobs of simpleSolution: 7
Tardy jobs of recommendedSolution: 8


Erwarteter Output:

    Tardy Jobs of simpleSolution: 7
    Tardy Jobs of recommendedSolution: 8

#### Aufgabe 2 - Zulässigkeitsprüfung (4 Punkte)

Bislang hatten Sie keinen Grund, an der richtigen Lösung zu zweifeln. Doch gestern Nacht kam es in der Abteilungfertigungsplanung zu einem Hackerangriff. Es kann sein, dass einige Planungsdaten beschädigt bzw. zerstört wurden.

Schreiben Sie eine Funktion *ValidateSolution*, die eine Lösung auf Ihre Zulässigkeit hin überprüft. Dabei sollten folgende Fälle betrachtet werden:
- In einer Lösungsrepräsentation sind zu wenige/ zu viele Auftrags-Ids enthalten.
- In einer Lösungsrepräsentation sind Auftrags-Ids doppelt enthalten.
- Die Start- und Endzeiten eines Auftrags auf aufeinanderfolgenden Maschinen überschneiden sich.

Erstellen Sie für jeden Fall geeignete Ausgaben. Die Funktion *ValidateSolution* sollte **alle** Fehler einer Lösung ausgeben. Wenn kein Fehler gefunden werden konnte, soll dies ebenfalls ausgegeben werden.

In [4]:
# Schreiben Sie hier Ihren Programmcode
def ValidateSolution(solution):
    intersect = False
    title = 'Reported Corruptions: \n' ##
    # Check für den ersten Corruption-Fall: 1 Punkt
    ## Fehler anfällig, da nicht immer 10 Jobs!!
    if len(solution.Permutation) != 10:
        title += '\t- Detected wrong number of IDs in solution representation.\n'
    # Check für den zweiten Corruption-Fall: 1 Punkt
    if len(solution.Permutation) != len(set(solution.Permutation)):
        title += '\t- Detected duplicate IDs in solution representation.\n'
    # Check für den dritten Corruption-Fall: 1 Punkt
    if intersect == False:
        for job in solution.OutputJobs.values():
             for i in range(1, len(job.EndTimes)):
                if job.EndTimes[i] < job.StartTimes[i - 1]:
                    title += '\t- Die Start- und Endzeiten eines Auftrags auf aufeinanderfolgenden Maschinen überschneiden sich.\n'
                    intersect == True
                    break

    if (len(solution.Permutation) == 10) and (len(solution.Permutation) == len(set(solution.Permutation))) and intersect == False:
        title += '\t- No corruptions detected.\n'

    return title
        



# Führen Sie im Anschluss den folgenden Code aus
print(ValidateSolution(simpleSolution))   
print(ValidateSolution(Solution(data.InputJobs, [0, 3, 2, 3, 1, 4, 5, 8])))
print(ValidateSolution(Solution(data.InputJobs, [0, 3, 2, 1, 4, 5, 8, 7, 10, 9, 11, 12])))

Reported Corruptions: 
	- No corruptions detected.

Reported Corruptions: 
	- Detected wrong number of IDs in solution representation.
	- Detected duplicate IDs in solution representation.

Reported Corruptions: 
	- Detected wrong number of IDs in solution representation.



In [15]:
def ValidateSol(dataSet, solution):

    title = 'Reported Corruptions: \n'
    if len(solution.Permutation) != len(dataSet.InputJobs):
        title += '\t - Detected wrong number of IDs in solution representation.\n'
    if len(set(solution.Permutation)) != len(solution.Permutation):
        title += '\t - Detected duplicate IDs in solution representation.\n'
    if True:
        for job in solution.OutputJobs.values():
            for i in range(len(job.EndTimes)):
                if job.EndTimes[i] > job.StartTimes[i+1]:
                    title += '\t Detected intersected start- and endtimes. \n'
                    
    if title == 'Reported Corruptions: \n':
        title += '\t - No corruptions detected.\n'

print(ValidateSolution(simpleSolution))   
print(ValidateSolution(Solution(data.InputJobs, [0, 3, 2, 3, 1, 4, 5, 8])))
print(ValidateSolution(Solution(data.InputJobs, [0, 3, 2, 1, 4, 5, 8, 7, 10, 9, 11, 12])))


Reported Corruptions: 
	- No corruptions detected.

Reported Corruptions: 
	- Detected wrong number of IDs in solution representation.
	- Detected duplicate IDs in solution representation.

Reported Corruptions: 
	- Detected wrong number of IDs in solution representation.



In [ ]:

def ValidateSolution(solution):
    intersect = False
    title = 'Reported Corruptions: \n' ##


    if len(solution.Permutation) != 10:
        title += '\t- Detected wrong number of IDs in solution representation.\n'

    if len(solution.Permutation) != len(set(solution.Permutation)):
        title += '\t- Detected duplicate IDs in solution representation.\n'

    if intersect == False:
        for job in solution.OutputJobs.values():
             for i in range(1, len(job.EndTimes)):
                if job.EndTimes[i] < job.StartTimes[i - 1]:
                    title += '\t- Die Start- und Endzeiten eines Auftrags auf aufeinanderfolgenden Maschinen überschneiden sich.\n'
                    intersect == True
                    break

    if (len(solution.Permutation) == 10) and (len(solution.Permutation) == len(set(solution.Permutation))) and intersect == False:
        title += '\t- No corruptions detected.\n'

    return title

Erwarteter Output:

	Reported Corruptions: 
		- No corruptions detected.
	Reported Corruptions: 
		- Detected wrong number of IDs in solution representation.
		- Detected duplicate IDs in solution representation.

	Reported Corruptions: 
		- Detected wrong number of IDs in solution representation.

#### Aufgabe 3 - Die CDS Heuristik von Campell, Dudek & Smith (11 Punkte)

Neben der NEH Heuristik zählt der Ansatz von Campbell, Dudek und Smith zu den bekanntesten und besten konstruktiven Lösungsansätzen für das Flow Shop Problem. Das Verfahren basiert dabei auf der Idee des Johnson-Algorithmus. Da der Praktikant in der Fertigungsplanung bereits einiges über den Johnson-Algorithmus weiß, hat er sich auch an einer Implementierung der CDS Heuristik versucht. Leider funktioniert der Algorithmus jedoch noch nicht so wie erwartet.

Informieren Sie sich bitte über die CDS Heuristik mithilfe der beigefügten Literatur. Berichtigen Sie anschließend den Programmcode der Funktion **CampellDudekSmith()**, die der Praktikant bereits erstellt hat. Überprüfen Sie anschließend Ihre Korrektur durch den vergleich mit dem **erwarteten Output**. Bei der Umsetzung müssen **keine** Rüstzeiten beachtet werden.

Kommentieren Sie den Programmcode und weisen Sie die Fehler, die Sie entdeckt haben, explizit aus!

**Tipp:**
<details>
Nutzen Sie zur Kontrolle die Debug-Funktion in Python.
</details>

In [5]:
# Hier soll es eine Funktion geben. nicht class: 1 Punkt
def CampellDudekSmith(inputJobs):

    solutionPool = {}
    nMachines = len(inputJobs[0].Operations) 

    for i in range(nMachines-1): # i soll die Anzahl von Maschinen sein, hat nichts zu tun mit jobs. i muss < nMachines sein: 2 Punkte

        minimumFirstMachine = []
        notMinimumFirstMachine = []

        for j, job in enumerate(inputJobs):
            # Reihenfolge von theta sind korrigiert. : 1 Punkt
            thetaFirst = sum(job.ProcessingTime(x) for x in range(0,i + 1))
            thetaSecond = sum(job.ProcessingTime(x) for x in range(nMachines-i-1,nMachines)) # hier muss nMachines-i-1 sein, zB i == 0, gibt es kein x fuer thetasecond

            if(thetaFirst < thetaSecond): #wenn Zeit anf M1 weniger als M2: 1 Punkt
                minimumFirstMachine.append((j,thetaFirst))
            else:
                notMinimumFirstMachine.append((j,thetaSecond)) # thetaSecond gehoert zu notMinimumFirstMachine: 1 Punkt

        minimumFirstMachine.sort(key=lambda x: x[1]) # sorted by x[1]: 1 Punkt
        notMinimumFirstMachine.sort(key=lambda x: x[1], reverse = True)
     
        tmpPermutation = [x[0] for x in minimumFirstMachine]
        tmpPermutation.extend([x[0] for x in notMinimumFirstMachine])
  
        tmpSolution = Solution(inputJobs, tmpPermutation)
        EvaluationLogic().DefineStartEnd(tmpSolution)  # !!!!!!!: 1 Punkt
        solutionPool[i] = tmpSolution


    bestSolution = solutionPool[0]
    for sol in solutionPool.values():
        if bestSolution.Makespan > sol.Makespan:
            bestSolution = sol

    EvaluationLogic().DefineStartEnd(bestSolution) ## 2 Punkte
    return bestSolution
  

# man muss richtigen Name von Methode benutzen
# wir suchen nach Loesungen, dann InputJobs ist richtig: 1 Punkt
CDSSol = CampellDudekSmith(data.InputJobs)
print(CDSSol)


The permutation [5, 8, 3, 0, 6, 9, 4, 2, 7, 1] results in a Makespan of 1159


In [17]:

def CampellDudekSmith(inputJobs):

    solutionPool = {}
    nMachines = len(inputJobs[0].Operations) 

    for i in range(nMachines-1): 

        minimumFirstMachine = []
        notMinimumFirstMachine = []

        for j, job in enumerate(inputJobs):
           
            thetaFirst = sum(job.ProcessingTime(x) for x in range(0,i + 1))
            thetaSecond = sum(job.ProcessingTime(x) for x in range(nMachines-i-1,nMachines)) 

            if(thetaFirst < thetaSecond): 
                minimumFirstMachine.append((j,thetaFirst))
            else:
                notMinimumFirstMachine.append((j,thetaSecond)) 

        minimumFirstMachine.sort(key=lambda x: x[1]) 
        notMinimumFirstMachine.sort(key=lambda x: x[1], reverse = True)
     
        tmpPermutation = [x[0] for x in minimumFirstMachine]
        tmpPermutation.extend([x[0] for x in notMinimumFirstMachine])
  
        tmpSolution = Solution(inputJobs, tmpPermutation)
        EvaluationLogic().DefineStartEnd(tmpSolution)  
        solutionPool[i] = tmpSolution


    bestSolution = solutionPool[0]
    for sol in solutionPool.values():
        if bestSolution.Makespan > sol.Makespan:
            bestSolution = sol

    EvaluationLogic().DefineStartEnd(bestSolution) 
    return bestSolution
  

CDSSol = CampellDudekSmith(data.InputJobs)
print(CDSSol)

The permutation [5, 8, 3, 0, 6, 9, 4, 2, 7, 1] results in a Makespan of 1159


Erwarteter Output:
    
    The permutation [5, 8, 3, 0, 6, 9, 4, 2, 7, 1] results in a Makespan of 1159